source/auth/tests/connection-string.json

```diff
@@ -1,279 +1,344 @@
+      "auth": null,
+      "options": null
+    },
+    {
+      "description": "should recognize the mechanism (MONGODB-X509)",
+      "uri": "mongodb://CN%3DmyName%2COU%3DmyOrgUnit%2CO%3DmyOrg%2CL%3DmyLocality%2CST%3DmyState%2CC%3DmyCountry@localhost/?authMechanism=MONGODB-X509",
+      "hosts": null,
+      "valid": true,
+      "warning": false,
+      "auth": {
+        "username": "CN=myName,OU=myOrgUnit,O=myOrg,L=myLocality,ST=myState,C=myCountry",
+        "password": null,
+        "db": "$external"
+      },
+      "options": {
+        "authmechanism": "MONGODB-X509"
+      }
+    },
+    {
+      "description": "should recognize the mechanism where authSource $external (MONGODB-X509)",
+      "uri": "mongodb://CN%3DmyName%2COU%3DmyOrgUnit%2CO%3DmyOrg%2CL%3DmyLocality%2CST%3DmyState%2CC%3DmyCountry@localhost/foo?authMechanism=MONGODB-X509&authSource=$external",
+      "hosts": null,
+      "valid": true,
+      "warning": false,
+      "auth": {
+        "username": "CN=myName,OU=myOrgUnit,O=myOrg,L=myLocality,ST=myState,C=myCountry",
+        "password": null,
+        "db": "$external"
+      },
+      "options": {
+        "authmechanism": "MONGODB-X509"
+      }
+    },
+    {
+      "description": "should throw an exception when an authMechanism (MONGODB-X509) has an authSource other than $external",
+      "uri": "mongodb://CN%3DmyName%2COU%3DmyOrgUnit%2CO%3DmyOrg%2CL%3DmyLocality%2CST%3DmyState%2CC%3DmyCountry@localhost/foo?authMechanism=MONGODB-X509&authSource=bar",
+      "hosts": null,
+      "valid": false,
+      "warning": true,
+      "auth": null,
+      "options": null
+    },
+    {
+      "description": "should recognize the mechanism (MONGODB-X509) with no username",
+      "uri": "mongodb://localhost/?authMechanism=MONGODB-X509",
+      "hosts": null,
+      "valid": true,
+      "warning": false,
+      "auth": {
+        "username": null,
+        "password": null,
+        "db": "$external"
+      },
+      "options": {
+        "authmechanism": "MONGODB-X509"
+      }
+    },
+    {
+      "description": "should not support passwords with the mechanism (MONGODB-X509)",
+      "uri": "mongodb://user:password@localhost/?authMechanism=MONGODB-X509",
+      "hosts": null,
+      "valid": false,
+      "warning": false,
+      "auth": null,
+      "options": null
+    },
+    {
+      "description": "should throw an exception if provided with an invalid authSource (MONGODB-X509)",
+      "uri": "mongodb://user@localhost/?authMechanism=MONGODB-X509&authSource=foo",
+      "hosts": null,
+      "valid": false,
+      "warning": false,
+      "auth": null,
+      "options": null
+    },
+    {
+      "description": "should recognize the mechanism (PLAIN)",
+      "uri": "mongodb://user:password@localhost/?authMechanism=PLAIN",
+      "hosts": null,
+      "valid": true,
+      "warning": false,
+      "auth": {
+        "username": "user",
+        "password": "password",
+        "db": "$external"
```

Comment:
I don't think this is right. PLAIN can be used with things other than LDAP, so doesn't technically require $external.

Reply:
According to Spec: _**Source** MUST be $external_

https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#plain

Reply:
Looking at the docs and I'm confused if it is valid to be something other than `$external`:
 
* [LDAP Auth via saslauthd](https://docs.mongodb.com/manual/core/security-ldap/#ldap-authentication-via-saslauthd) reads as if it should be `$external`.
* [Security LDAP external](https://docs.mongodb.com/manual/core/security-ldap-external/#mongodb-roles-for-ldap-authorization) isn't clear to me as either way.

Can you confirm if the Spec is correct or wrong? and I'll action accordingly.

Reply:
```
$ ./mongod --dbpath ~/data/db --auth --setParameter authenticationMechanisms=PLAIN,SCRAM-SHA-1,SCRAM-SHA-256 -vv
```

```
>>> c.admin.authenticate("user", "password", mechanism='PLAIN')
True
```

```
2018-04-11T15:25:51.076-0700 I ACCESS   [conn2] Successfully authenticated as principal user on admin
2018-04-11T15:25:51.076-0700 I COMMAND  [conn2] command admin.system.users command: saslStart { saslStart: 1, mechanism: "PLAIN", payload: "xxx", autoAuthorize: 1, $db: "admin" } numYields:0 reslen:94 locks:{ Global: { acquireCount: { r: 4 } }, Database: { acquireCount: { r: 2 } }, Collection: { acquireCount: { r: 2 } } } protocol:op_query 17ms

```

Reply:
So, the spec is wrong. With PLAIN you're delegating the task of authentication to the server itself. The PLAIN mechanism is so named because you send the plain text username and password to the server.

https://tools.ietf.org/html/rfc4616